Svm benchmark

PhMueller opened the pull request:
SVM Benchmark from hpolib1.5.

KEggensperger commented:
This PR is now ready for reviewing
@aaronkl @mfeurer

mfeurer commented on .travis.yml:
```diff
@@ -33,8 +33,8 @@ install:
   # Install auto-sklearn
-  - curl https://raw.githubusercontent.com/automl/auto-sklearn/development/requirements.txt | xargs -n 1 -L 1 pip install
-  - pip install git+https://github.com/automl/auto-sklearn@development
+  #- curl https://raw.githubusercontent.com/automl/auto-sklearn/development/requirements.txt | xargs -n 1 -L 1 pip install
```
Does this imply that `auto-sklearn` should be an optional dependency?

KEggensperger replied:
No, auto-sklearn is installed in l33 as a mandatory requirement

mfeurer commented on examples/SMAC3_on_LogisticRegression10CVOnMnist.py:
```diff
@@ -0,0 +1,54 @@
+
+logReg = logistic_regression.LogisticRegression10CVOnMnist(rng=10)
+
+logger = logging.getLogger("Optimizer")  # Enable to show Debug outputs
```
I don't see why one needs the `logger` variable.

KEggensperger replied:
Agreed and removed

mfeurer commented on examples/SMAC3_on_LogisticRegression10CVOnMnist.py:
```diff
@@ -0,0 +1,54 @@
+                     "cs": cs,
+                     "deterministic": "true",
+                     "instances": list([str(i) for i in range(10)]), # train_inst_fh.name,
+                     "test_instances": ["10", ]
```
You don't need the comma in a list.

mfeurer commented on examples/SMAC3_on_LogisticRegression10CVOnMnist.py:
```diff
@@ -0,0 +1,54 @@
+                     "test_instances": ["10", ]
+                     })
+
+# Evaluate default config as a test
```
I don't understand the comment.

KEggensperger replied:
Agreed and removed

mfeurer commented on examples/SMAC3_on_LogisticRegression10CVOnMnist.py:
```diff
@@ -0,0 +1,54 @@
+    incumbent = smac.solver.incumbent
+
+# Then evaluate final configuration
+inc_value = taf.run(incumbent)[1]
```
Shouldn't that run on the test set?

KEggensperger replied:
Agreed and changed

mfeurer commented on hpolib/benchmarks/ml/logistic_regression.py:
```diff
@@ -4,46 +4,55 @@
     @AbstractBenchmark._configuration_as_array
     def objective_function(self, x, dataset_fraction=1, **kwargs):
+        # Shuffle training data
+        rng = kwargs.get("rng", None)
```
`None` is the default return value of `dict.get()` and not necessary.

mfeurer commented on hpolib/benchmarks/ml/logistic_regression.py:
```diff
@@ -53,51 +62,41 @@ def objective_function(self, x, dataset_fraction=1, **kwargs):
-        l2_reg = np.float32(x[1])
-        batch_size = np.int32(x[2])
-        dropout_rate = np.float32(x[3])
+        # if rng was not not, set rng for lasagne
```
two `not` in the comment

KEggensperger replied:
Agreed

mfeurer commented on hpolib/benchmarks/ml/logistic_regression.py:
```diff
@@ -120,6 +119,32 @@ def get_meta_information():
+    def _train_model(self, config, train, train_targets, valid, valid_targets):
+        """ helper method that accepts data and configuration and returns
+        lc_curve, cost_curve
+        :param config: list
```
shouldn't this be in numpy format?

KEggensperger replied:
Not my code, but changed in anyway

mfeurer commented on hpolib/benchmarks/ml/logistic_regression.py:
```diff
@@ -120,6 +119,32 @@ def get_meta_information():
+        :param valid_targets: list
+        :return: lc_curve, cost_curve
+        """
+        learning_rate = float(10 ** config[0])
```
I assume there was a reason why the values were casted to `np.float32` before this change request.

KEggensperger replied:
I don't know. @aaronkl: Is there a reason for this?

mfeurer commented on hpolib/util/rng_helper.py:
```diff
@@ -0,0 +1,36 @@
+    if rng is None returns self_rng if RandomState
+    if self_rng is None initializes RandomState at random
+
+    :param rng: int or RandomState
```
Shouldn't this be in numpy docstyle format?

KEggensperger replied:
Agreed and changed it

mfeurer commented on hpolib/util/rng_helper.py:
```diff
@@ -0,0 +1,36 @@
+
+def create_rng(rng):
+    """ helper to create rng from RandomState or int
+    :param rng: int or RandomState
```
Shouldn't this be in numpy docstyle format?

mfeurer commented on setup.py:
```diff
@@ -4,19 +4,12 @@
     version=version,
     packages=setuptools.find_packages(exclude=['test']),
-    install_requires=requirements,
+    install_requires=[],
```
why does this PR remove the requirements?

KEggensperger replied:
Because the documentation recommends to install them independently and our current requirements.txt contains comments which breaks our current setup.py

PhMueller committed:
ADD SVM Benchmark

PhMueller committed:
ADD SVM Benchmark Container

PhMueller committed:
PEP8 + FLAKE8

KEggensperger commented on extra_requirements/svm.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "svm": ["pandas>=0.22.2,<0.24.2","openml==0.10.2"]
```
shouldn't this also require sklearn?

PhMueller replied:
yes, you are right! 
I removed it first, because openml requires scikit-learn>=0.18.1 already. But yeah: Explicit > Implicit.

KEggensperger commented on hpolib/benchmarks/ml/svm_benchmark.py:
```diff
@@ -0,0 +1,316 @@
+            self.shuffle_data(self.rng)
+
+        # Split of dataset subset
+        train_size = max(fidelity['dataset_fraction'],
```
Can your raise a warning if not the fraction given by the fidelity is used?

KEggensperger commented on hpolib/benchmarks/ml/svm_benchmark.py:
```diff
@@ -0,0 +1,316 @@
+                                         replace=False)
+
+        # Use 10 time the number of classes as lower bound for the dataset fraction
+        n_classes = np.unique(self.y_train).shape[0]
```
If this a heuristic you came up with or can we refer to some work for this?

PhMueller replied:
Aaron used it in his paper: 

[Fast Bayesian Optimization of Machine Learning Hyperparameters
on Large Datasets](https://arxiv.org/pdf/1605.07079.pdf)
> We implemented Hyperband following Li et al. (2017) using
the recommended setting for the parameter η = 3 that controls the intermediate subset sizes. For each experiment, we
adjusted the budget allocated to each Hyperband iteration to
allow the same minimum dataset size as for FABOLAS: 10
times the number of classes for the support vector machine
benchmarks and the maximum batch size for the neural
network benchmarks

KEggensperger replied:
Sounds good, can you add a comment referencing that paper?

KEggensperger commented on hpolib/container/recipes/ml/Singularity.SupportVectorMachine:
```diff
@@ -0,0 +1,21 @@
+    && git clone https://github.com/automl/HPOlib2.git \
+    && cd HPOlib2 \
+    && git checkout master \
+    && pip install .[singularity,svm] \
```
the singularity requirements do no longer exist, correct?

KEggensperger commented on hpolib/benchmarks/ml/svm_benchmark.py:
```diff
@@ -0,0 +1,316 @@
+                               "author = {A. Klein and S. Falkner and S. Bartels and P. Hennig and F. Hutter},"
+                               "title = {Fast Bayesian Optimization of Machine Learning"
+                               "Hyperparameters on Large Datasets},"
+                               "journal = corr,"
```
This is out of date. The paper has been published. Please check our publication website (https://ml.informatik.uni-freiburg.de/publications.html) for the papers in 2017 and double-check whether any of these used this non-surrogate based version?

KEggensperger commented on hpolib/benchmarks/ml/svm_benchmark.py:
```diff
@@ -0,0 +1,316 @@
+                 ("categorical", OneHotEncoder(categories=self.categories, sparse=False), self.categorical_data),
+                 ("continuous", "passthrough", ~self.categorical_data)])),
+            ('svm',
+             svm.SVC(gamma=gamma, C=C, random_state=self.rng))
```
Could you please add the cache size here as well and make this a class attribute set to the default value so one could easily change this from the outside.

KEggensperger commented on hpolib/benchmarks/ml/svm_benchmark.py:
```diff
@@ -0,0 +1,316 @@
+        nan_columns = np.all(np.isnan(self.X_train), axis=0)
+        self.categorical_data = self.categorical_data[~nan_columns]
+
+        mean_imputer = SimpleImputer(strategy='mean')
```
This should be part of the pipeline as well and only done on the given fraction of the training data

PhMueller committed:
[skip travis] Update Requirements for svm and xgb

PhMueller committed:
[skip travis] Fix wrong dependency in recipe

PhMueller committed:
[skip travis] Move the imputer into objective function in xgboost_benchmark.py and svm_benchmark.py

- Move the imputer into objective function in xgboost_benchmark.py and svm_benchmark.py
- Update svm references

PhMueller committed:
[skip travis] Move the imputer to the pipeline

- Move the imputer to the pipeline
  This requires however that not NaN value is in the categorical data columns.
  Impute them by using a unsued value (here: the smalles value in this column - 1)
  They have to be numerical, since all openml values are numerical and sklearn doesn't allow to mix scalar and strings.
- Impute in a first step only the categorical then transfrom them to one hot
- Use the mean imputer only for the continuous feature
- Fix the same behaviiour in the xgboost benchmark

PhMueller committed:
[skip travis] Flake + Pep8

PhMueller committed:
Tests

PhMueller committed:
[skip travis] Fix error in imputer args

PhMueller committed:
[skip travis] Update Test

PhMueller committed:
Update in the client interface

- Receiving the fidelity space did not set the seed.
